### PR TITLE
Delivery to target warehouse

### PR DESCRIPTION
**Pull request by nabinhait**
Against #3970

### CODE REVIEW

File: erpnext/stock/doctype/delivery_note/test_delivery_note.py [hunk 1]
```diff
@@ -321,12 +319,58 @@ def test_return_for_serialized_items(self):
 		dn.cancel()
 		
 		self.check_serial_no_values(serial_no, {
-			"status": "Available",
 			"warehouse": "_Test Warehouse - _TC",
 			"delivery_document_no": "",
 			"purchase_document_no": se.name
 		})
+		
+	def test_delivery_of_bundled_items_to_target_warehouse(self):
```
Comment (anandpdoshi) [inline erpnext/stock/doctype/delivery_note/test_delivery_note.py]: In this test, we have passed the target warehouse as `_Test Warehouse 1 - _TC`, but we haven't changed the target warehouse in bundled item to something different. Isn't that the requirement of #3970 ?

File: erpnext/stock/doctype/serial_no/serial_no.py [hunk 2]
```diff
@@ -162,10 +139,10 @@ def get_stock_ledger_entries(self):
 		return sle_dict
 
 	def on_trash(self):
-		if self.status == 'Delivered':
-			frappe.throw(_("Delivered Serial No {0} cannot be deleted").format(self.name))
 		if self.warehouse:
 			frappe.throw(_("Cannot delete Serial No {0} in stock. First remove from stock, then delete.").format(self.name))
+		elif self.delivery_document_no:
+			frappe.throw(_("Delivered Serial No {0} cannot be deleted").format(self.name))
```
Comment (anandpdoshi) [inline erpnext/stock/doctype/serial_no/serial_no.py]: Instead, shouldn't we check if there is any Stock Ledger Entry for this serial no?

File: erpnext/stock/doctype/delivery_note/test_delivery_note.py (hunk 1 above)
Comment (nabinhait) [inline erpnext/stock/doctype/delivery_note/test_delivery_note.py]: Target warehouse get copied from main item table, if not specified in packing list

File: erpnext/stock/doctype/serial_no/serial_no.py (hunk 2 above)
Comment (nabinhait) [inline erpnext/stock/doctype/serial_no/serial_no.py]: Fixed

**Comment by anandpdoshi** [comment]
@nabinhait test case is failing

**Comment by rmehta** [comment]
Add this:

- [ ] Ability to change Serial No Item in Manufacture / Repack type of operation

**Comment by rmehta** [comment]
@sliska This is due in next release. Sorry for the delay (we were tied up with conference)

**Comment by sliska** [comment]
Hi Rushabh,

When will you have time for the second deliverable? (Buy Product Bundles
(#3969))

We had agreed to push it back until October. Will it be ready for the next
release?

thanks,
Sofia

**Pull request merged**